<a href="https://colab.research.google.com/github/avawolsborn/Ava-Wolsborn-DS1002/blob/main/06_Pandas_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pandas exercises 1

1. Import dependencies

In [2]:
import pandas as pd

---

In [3]:
list = [98, 24, 46, 87, 33]

2. Convert `list` to a pandas series using `pd.Series`

In [4]:
list = pd.Series(list)

3. Take the above series and add an index that starts at 2 and steps by 3

In [5]:
list.index = [2, 5, 8, 11, 14]

list

2     98
5     24
8     46
11    87
14    33
dtype: int64

4. Create a dictionary of 4 names, ages, and careers   
Convert this dictionary to a pandas dataframe: `df`.  
Print the dataframe.

In [6]:
df = pd.DataFrame({
    'Name': ['Riley', 'Tara', 'Ava', 'Tom'],
    'Age': [19, 18, 18, 5],
    'Career': ['Doctor', 'Data Scientist', 'Pilot', 'Astronaut'],
})

df

Name  Age          Career
0  Riley   19          Doctor
1   Tara   18  Data Scientist
2    Ava   18           Pilot
3    Tom    5       Astronaut

5. Name the index of `df` 'participant_id'.  
Print `df`.

In [7]:
df.index.name = 'participant_id'

df

Name  Age          Career
participant_id                            
0               Riley   19          Doctor
1                Tara   18  Data Scientist
2                 Ava   18           Pilot
3                 Tom    5       Astronaut

6. Use a pandas function to determine the data types of the columns of `df`.

In [8]:
df.dtypes

Name      object
Age        int64
Career    object
dtype: object

7. Using bracket notation, reference column `name`

In [9]:
df['Name']

participant_id
0    Riley
1     Tara
2      Ava
3      Tom
Name: Name, dtype: object

8. Using dot notation, reference column `name`

In [10]:
df.Name

participant_id
0    Riley
1     Tara
2      Ava
3      Tom
Name: Name, dtype: object

8. Reference column `name` and have it returned in a dataframe.

In [11]:
pd.DataFrame(df.Name)

Name
participant_id       
0               Riley
1                Tara
2                 Ava
3                 Tom

9. Show the 3rd element in column `career`.

In [12]:
df.Career.values[2]

'Pilot'

10.  Load the `iris` dataset that we used in class.  Use `iris` as the name of your data frame.

In [13]:
import seaborn as sns
iris = sns.load_dataset('iris')

11.  What is the correlation between the variables `sepal_length` and `sepal_width` for all the irises?

In [14]:
iris[['sepal_length','sepal_width']].corr()

sepal_length  sepal_width
sepal_length       1.00000     -0.11757
sepal_width       -0.11757      1.00000

There is a correlation of -0.11757 between the variables sepal_length and sepal_width.

12.  Which of the three species has the highest correlation between `sepal_length` and `sepal_width`?  Show any calculations you use to support your answer.

In [16]:
## For each species:
# define a species-specific data frame that prints the information from iris of only that species
# print the species name and the correlation values of the two variables for that specific species: sepal_length and sepal_width

setosa_species = iris[iris.species == 'setosa']
print("Setosa:")
print(setosa_species[['sepal_length','sepal_width']].corr())

versicolor_species = iris[iris.species == 'versicolor']
print("Versicolor:")
print(versicolor_species[['sepal_length','sepal_width']].corr())

virginica_species = iris[iris.species =='virginica']
print("Virginica:")
print(virginica_species[['sepal_length','sepal_width']].corr())

Setosa:
              sepal_length  sepal_width
sepal_length      1.000000     0.742547
sepal_width       0.742547     1.000000
Versicolor:
              sepal_length  sepal_width
sepal_length      1.000000     0.525911
sepal_width       0.525911     1.000000
Virginica:
              sepal_length  sepal_width
sepal_length      1.000000     0.457228
sepal_width       0.457228     1.000000


The setosa species has the highest correlation between the variables sepal_length and sepal_width.

13.  Among the four variables, the two that are most negatively correlated (overall) are `sepal_width` and `petal_length`.  (How do we know this?)  Determine the same thing but for each species separately:  for each species, which two variables have the most negative (or the least positive) correlation?  Again, include calculations to support your answer.

In [17]:
# Repeat the same process as the previous code chunk but with every single variable instead

print("Setosa:")
print(setosa_species[['sepal_length','sepal_width','petal_length','petal_width']].corr())

print("Versicolor:")
print(versicolor_species[['sepal_length','sepal_width','petal_length','petal_width']].corr())

print("Virginica:")
print(virginica_species[['sepal_length','sepal_width','petal_length','petal_width']].corr())

Setosa:
              sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000     0.742547      0.267176     0.278098
sepal_width       0.742547     1.000000      0.177700     0.232752
petal_length      0.267176     0.177700      1.000000     0.331630
petal_width       0.278098     0.232752      0.331630     1.000000
Versicolor:
              sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000     0.525911      0.754049     0.546461
sepal_width       0.525911     1.000000      0.560522     0.663999
petal_length      0.754049     0.560522      1.000000     0.786668
petal_width       0.546461     0.663999      0.786668     1.000000
Virginica:
              sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000     0.457228      0.864225     0.281108
sepal_width       0.457228     1.000000      0.401045     0.537728
petal_length      0.864225     0.401045      1.000000     0.322108
petal_width       0.281108     

The variables with the least positive correlation for each species:

Setosa: petal_length & sepal_width

Versicolor: sepal_width & sepal_length

Virginica: petal_width & sepal_length

14.  Once again, the variables `sepal_width` and `petal_length` are negatively correlated.  But for each individual species, these two variables are positively correlated.  How is this possible?  Explain this phenomenon.  Use a data visualization if it helps.


This is possible because with each individual species, as the sepal length increases, the sepal width generally does as well. However, every species holds a different average sepal length and width. For example, the sepal length of virginicas is greater than that of setosas, but the average sepal width is less than that of setosas. As there is an increase in the average sepal length between species, there is a decrease in the average sepal width for each.

With that, we find a data visualization that will appear to be increasing in 3 separate sections but decreasing in values when looking across all 3 species. As a result, we end up with an overall negative correlation and a positive correlation within each species.